src/fu-noinst.c

```diff
@@ -0,0 +1,606 @@
+		     NULL,
+		     /* TRANSLATORS: command description */
+		     _("Get all devices that support firmware updates"),
+		     fu_util_get_devices);
+	fu_util_add (priv->cmd_array,
+		     "watch",
+		     NULL,
+		     /* TRANSLATORS: command description */
+		     _("Watch for hardare changes"),
+		     fu_util_watch);
+	fu_util_add (priv->cmd_array,
+		     "install-blob",
+		     "FILENAME DEVICE-ID",
+		     /* TRANSLATORS: command description */
+		     _("Install a firmware blob on a device"),
+		     fu_util_install_blob);
+	fu_util_add (priv->cmd_array,
+		     "attach",
+		     "DEVICE-ID",
+		     /* TRANSLATORS: command description */
+		     _("Attach to firmware mode"),
+		     fu_util_attach);
+	fu_util_add (priv->cmd_array,
+		     "detach",
+		     "DEVICE-ID",
+		     /* TRANSLATORS: command description */
+		     _("Detach to bootloader mode"),
+		     fu_util_detach);
+
+	/* do stuff on ctrl+c */
+	priv->cancellable = g_cancellable_new ();
+	g_unix_signal_add_full (G_PRIORITY_DEFAULT,
+				SIGINT, fu_util_sigint_cb,
+				priv, NULL);
+	g_signal_connect (priv->cancellable, "cancelled",
+			  G_CALLBACK (fu_util_cancelled_cb), priv);
+
+	/* sort by command name */
+	g_ptr_array_sort (priv->cmd_array,
+			  (GCompareFunc) fu_sort_command_name_cb);
+
+	/* get a list of the commands */
+	priv->context = g_option_context_new (NULL);
+	cmd_descriptions = fu_util_get_descriptions (priv->cmd_array);
+	g_option_context_set_summary (priv->context, cmd_descriptions);
+	g_option_context_set_description (priv->context,
+		"This tool allows an administrator to use the fwupd plugins "
+		"without being installed on the host system.");
+
+	/* TRANSLATORS: program name */
+	g_set_application_name (_("Firmware Utility"));
+	g_option_context_add_main_entries (priv->context, options, NULL);
+	ret = g_option_context_parse (priv->context, &argc, &argv, &error);
+	if (!ret) {
+		/* TRANSLATORS: the user didn't read the man page */
+		g_print ("%s: %s\n", _("Failed to parse arguments"),
+			 error->message);
+		return EXIT_FAILURE;
+	}
+
+	/* set verbose? */
+	if (verbose) {
+		g_setenv ("G_MESSAGES_DEBUG", "all", FALSE);
+	} else {
+		g_log_set_handler (G_LOG_DOMAIN, G_LOG_LEVEL_DEBUG,
+				   fu_util_ignore_cb, NULL);
+	}
+
+	/* load engine */
+	priv->engine = fu_engine_new ();
+	g_signal_connect (priv->engine, "device-added",
+			  G_CALLBACK (fu_main_engine_device_added_cb),
+			  priv);
+	g_signal_connect (priv->engine, "device-removed",
+			  G_CALLBACK (fu_main_engine_device_removed_cb),
+			  priv);
+	g_signal_connect (priv->engine, "device-changed",
+			  G_CALLBACK (fu_main_engine_device_changed_cb),
+			  priv);
+	g_signal_connect (priv->engine, "status-changed",
+			  G_CALLBACK (fu_main_engine_status_changed_cb),
+			  priv);
+	g_signal_connect (priv->engine, "percentage-changed",
+			  G_CALLBACK (fu_main_engine_percentage_changed_cb),
+			  priv);
+	//FIXME: do we need to tell the engine the runtime AND build-time PLUGINDIR location?
```

Comment:
Can you check for lack of arguments before loading the engine?  It should speed up the error scenario of no command added.

Reply:
Fixed.